src/main.rs

```diff
@@ -1,3 +1,204 @@
+
+                        let new_end_byte = start_byte + content_change.text.len();
+                        let input_edit = InputEdit {
+                            start_byte,
+                            old_end_byte,
+                            new_end_byte,
+                            start_position: position_to_point(range.start),
+                            old_end_position: position_to_point(range.end),
+                            new_end_position: byte_offset_to_point(
+                                &file_state.contents,
+                                new_end_byte,
+                            ),
+                        };
+                        file_state.tree.edit(&input_edit);
+                        file_state.tree = parse(&file_state.contents, Some(&file_state.tree));
+                    }
+                    None => {
+                        file_state.contents = (&*content_change.text).into();
+                        file_state.tree = parse_from_scratch(&file_state.contents);
+                    }
+                }
+            }
+        }
+
+        self.run_linting_and_report_diagnostics(&params.text_document.uri)
+            .await;
+    }
+}
+
+#[derive(Debug)]
+struct PerFileState {
+    contents: Rope,
+    tree: Tree,
+}
+
+fn parse_from_scratch(contents: &Rope) -> Tree {
+    parse(contents, None)
+}
+
+fn parse(contents: &Rope, old_tree: Option<&Tree>) -> Tree {
+    let mut parser = Parser::new();
+    parser
+        .set_language(SupportedLanguage::Rust.language())
+        .unwrap();
+    contents.parse(&mut parser, old_tree).unwrap()
+}
+
+fn lsp_position_to_char_offset(file_contents: &Rope, position: Position) -> usize {
+    file_contents.line_to_char(position.line as usize) + position.character as usize
+}
+
+fn position_to_point(position: Position) -> Point {
+    Point {
+        row: position.line as usize,
+        column: position.character as usize,
+    }
+}
+
+fn point_to_position(point: Point) -> Position {
+    Position {
+        line: point.row as u32,
+        character: point.column as u32,
+    }
+}
+
+fn byte_offset_to_point(file_contents: &Rope, byte_offset: usize) -> Point {
+    let line_num = file_contents.byte_to_line(byte_offset);
+    let start_of_line_byte_offset = file_contents.line_to_byte(line_num);
+    Point {
+        row: line_num,
+        column: byte_offset - start_of_line_byte_offset,
+    }
+}
+
+fn tree_sitter_range_to_lsp_range(file_contents: &Rope, range: tree_sitter::Range) -> Range {
+    Range {
+        start: point_to_position(byte_offset_to_point(file_contents, range.start_byte)),
+        end: point_to_position(byte_offset_to_point(file_contents, range.end_byte)),
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let stdin = tokio::io::stdin();
+    let stdout = tokio::io::stdout();
+
+    let (service, socket) = LspService::new(Backend::new);
+    Server::new(stdin, stdout, socket).serve(service).await;
```

Comment:
Most of the boilerplate (which there really isn't that much of, this was pretty straightforward/"clean") I got from either the `tower-lsp` docs or an example project or two